JBMC: Add symex-based program loading

smowton opened the pull request:
This adds the final few pieces required and enables symex-driven program loading in JBMC, controlled by the `--symex-driven-lazy-loading` command-line option. In this mode, functions are effectively finalised one-by-one, including committing to final location numbers, loop numbers and so on, so that references to instructions taken by symex remain valid in the eyes of subsequent passes.

Some JBMC options are currently incompatible: some for fundamental reasons (`--lazy-loading` specifies an alternate, incompatible loading scheme; `--all-properties` requests a list of assertions with no particular entry-point in mind and therefore nowhere to symex from; `--full-slice` fundamentally requires a whole program available), and one that will require a little extra work (`java-unwind-static-init` is currently implemented in a way that clashes with symex-driven loading, but can be reworked in a subsequent PR).

The whole JBMC test corpus is used to test symex-driven loading, with a small number of tests tagged to indicate incompatibility.

smowton committed:
Remove exceptions: add per-function entry point

This removes high-level exception-related instructions from a single function,
at the cost of some accuracy as we can no longer analyse other functions to
determine whether or not they may throw, and therefore always assume they will.

smowton committed:
Add tests for symex-driven lazy loading

This adds variants of the three test directories that use JBMC (cbmc-java, jbmc-strings and
strings-smoke-tests) to run with --symex-driven-lazy-loading, except for the small number of
tests that are expected to fail for harmless reasons, and one that is expected to fail and
which may be a bug (strings-smoke-tests/java_append_char).

smowton commented:
@NathanJPhillips please review

tautschnig commented:
Is it show-properties rather than all-properties that’s in conflict?

NathanJPhillips commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -499,22 +498,55 @@ void remove_exceptionst::operator()(goto_functionst &goto_functions)
   const namespacet ns(symbol_table);
   std::map<irep_idt, std::set<irep_idt>> exceptions_map;
   uncaught_exceptions(goto_functions, ns, exceptions_map);
+  // NOLINTNEXTLINE
```
Can the linter be fixed or this lint error ignored in review?

NathanJPhillips replied:
Couple of other instances of this too.

smowton replied:
Long-standing problem; haven't the time or inclination to dig into the linter just now...

NathanJPhillips commented on src/cbmc/bmc.cpp:
```diff
@@ -448,6 +388,107 @@ safety_checkert::resultt bmct::execute(const goto_functionst &goto_functions)
   }
 }
 
+safety_checkert::resultt bmct::execute(lazy_goto_modelt &lazy_model)
```
This parameter needs to be passed by r-value if you're going to `std::move` it below. i.e. the model needs to be moved into this function, indicating to the caller that it is no longer valid after this function has been called.

NathanJPhillips commented on src/jbmc/jbmc_parse_options.h:
```diff
@@ -89,14 +91,27 @@ class jbmc_parse_optionst:
+  // `const goto_functionst` (ahead-of-time function loading) or
+  // `lazy_goto_modelt` (symex-driven function loading).
+  template<class ModelorfunctionsT>
+    int run_bmc(
```
Is this normally indented?

NathanJPhillips reviewed (changes requested):
Most significant request is not to std::move something passed into a function by l-value.

tautschnig commented on regression/test.pl:
```diff
@@ -244,6 +266,8 @@ ($$$$)
   -D <key=value> Define - replace \$key string with "value" string in
                  test descriptors
+  -I <tag>   run only tests that have the given secondary tag. Can be repeated.
+  -X <tag>   exclude tests that have the given secondary tag. Can be repeated.
```
1. I think this commit (extending test.pl) could be moved to a separate PR
2. Actual change request: please extend the documentation that describes the syntax of test.desc files to cover the new tagging mechansim.

tautschnig replied:
Ignore the first one, but the second item still stands.

smowton replied:
Done in the other PR; will drop this commit here.

tautschnig commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -340,9 +340,13 @@ class goto_symex_statet final
   void switch_to_thread(unsigned t);
   bool record_events;
-  std::unique_ptr<const dirtyt> dirty;
+  std::unordered_set<irep_idt, irep_id_hash> dirty_processed_functions;
```
To me, this is the wrong place: this belong's to the state of the symbolic executor (`goto_symext`) and is not part of the symbolic program state.

tautschnig commented on src/cbmc/bmc.cpp:
```diff
@@ -355,72 +355,12 @@ void bmct::setup()
-    }
-
-    return decide(goto_functions, prop_conv);
+    return torun();
```
Why did you need to rewrite the entire function when seemingly just the type of the parameter should change?

smowton replied:
I needed to provide two alternate versions of `execute` (see below this function), which have different prefixes depending on whether they're using a lazy or an eager goto model, but then converge after symex. This particular factoring was to avoid duplicating the error-handling code between the two cases.

tautschnig commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -63,7 +63,7 @@ void goto_symext::symex_assign(
       statement == ID_java_new_array_data)
       symex_cpp_new(state, lhs, side_effect_expr);
-    else if(statement==ID_allocate)
+    else if(statement==ID_allocate || statement==ID_java_new)
```
How come? This should have been removed by `remove_java_new`!

smowton replied:
Previously remove_java_new could not be run per-function, because it lowered into a sized malloc, and types could later gain fields when method conversion encountered a field access against a stubbed type. However this is no longer true, so the whole remove_java_new malarkey can now go away again. I'll submit a PR to that effect on Monday.

tautschnig commented on regression/jbmc-strings/java_char_array_init/test.desc:
```diff
@@ -8,3 +8,9 @@ assertion.* file test_init.java line 33 .* SUCCESS$
+Runs indefinitely under symex-driven lazy loading. Unsure why yet; this could
+be a bug, or could simply be variations in order of VCCs / functions perturbing
+the string solver and leading to a long unfruitful search that would actually
+terminate eventually.
```
Please debug properly, for example by comparing the results of `--program-only`.

smowton replied:
These two have now been solved and therefore this change dropped.

tautschnig commented on regression/strings-smoke-tests/java_append_char/test.desc:
```diff
@@ -7,3 +7,8 @@ assertion.* file test_append_char.java line 23 .* SUCCESS
+Runs indefinitely under symex-driven lazy loading. Unsure why yet; this could
+be a bug, or could simply be variations in order of VCCs / functions perturbing
+the string solver and leading to a long unfruitful search that would actually
+terminate eventually.
```
As above: please debug. No throwing-over-the-fence, please.

smowton replied:
Change also dropped.

tautschnig reviewed (changes requested):
"Symex: handle java-new side-effect" seems to be a wrong fix for whatever-the-symptom as `ID_java_new` ought not to exist at this point. Also, the tests need to be debugged fully. Otherwise a few comments.

martin-cs commented on src/analyses/goto_check.h:
```diff
@@ -26,6 +26,7 @@ void goto_check(
 void goto_check(
   const namespacet &ns,
   const optionst &options,
+  const irep_idt &mode,
```
A string called "mode" is a bit leading ... no chance of a doxygen comment on what the inputs are is there?

martin-cs commented on src/goto-instrument/cover.h:
```diff
@@ -44,6 +55,14 @@ void instrument_cover_goals(
+  const cover_configt &,
+  goto_model_functiont &,
+  message_handlert &message_handler);
+
```
I guess this fits with @peterschrammel 's changes to coverage...

martin-cs reviewed (approved):
Looks fine from the point of view of the bits of code I look after, hence approve.  BUT there are some pretty wide API-breaking changes here so I'm hoping that someone with a wider view of that part of the code can comment.

smowton commented:
@tautschnig @NathanJPhillips changes applied, except as specifically noted. @tautschnig the dirty analysis check-if-function-already-analysed code has gone into `incremental_dirtyt`, defined alongside `dirtyt`.

@peterschrammel @thk123 @chrisr-diffblue this is now rebased and ready for review.

tautschnig commented on src/goto-symex/symex_main.cpp:
```diff
@@ -136,8 +136,17 @@ void goto_symext::symex_entry_point(
-  state.dirty=util_make_unique<dirtyt>(goto_functions);
-
+  state.dirty=util_make_unique<incremental_dirtyt>();
+  if(!pc->function.empty())
```
This ought to be an invariant. See https://github.com/diffblue/cbmc/pull/1799 for a discussion.

smowton replied:
Suggesting merging https://github.com/diffblue/cbmc/pull/1843 then it will be

smowton commented:
test.pl improvement moved to https://github.com/diffblue/cbmc/pull/1821 (but also still in this PR until that one lands).

tautschnig commented on src/goto-symex/symex_main.cpp:
```diff
@@ -168,22 +168,41 @@ void goto_symext::symex_threaded_step(
+static goto_symext::get_goto_functiont get_function_from_goto_functions(
+  const goto_functionst &goto_functions)
+{
+  /// NOLINTNEXTLINE
```
I don't think that's needed here, let alone as a doxygen comment

tautschnig commented on src/goto-symex/symex_main.cpp:
```diff
@@ -194,37 +213,62 @@ void goto_symext::operator()(
+  catch(const std::out_of_range &error)
+  {
     throw "the program has no entry point";
+  }
```
I'd suggest to turn this into a `PRECONDITION` - the problem can be detected much earlier and should detected by the caller. (I know this isn't really your bug, but since lots of this code is being touched here ... if you don't mind?)

smowton commented:
@tautschnig @karkhaz rebased this on top of the recent symex changes. Will peel off the first couple of commits as a separate PR later tonight.

smowton commented:
This has been rebased to exclude the commits that landed in #1844. I'm now expecting @karkhaz to rebase #1641, then this PR will be next in line after some rewriting of its changes to the BMC driver (cbmc/bmc.cpp).

NathanJPhillips commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -745,14 +847,56 @@ void jbmc_parse_optionst::process_goto_function(
+    if(using_symex_driven_loading)
+    {
+      if(show_loaded_functions(goto_model))
+        exit(0);
```
This may abort without flushing JSON output.

NathanJPhillips commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -475,26 +491,73 @@ int jbmc_parse_optionst::doit()
+    // unwinds <clinit> loops to number of enum elements
+    // side effect: add this as explicit unwind to unwind set
+    if(options.get_bool_option("java-unwind-enum-static"))
+      remove_static_init_loops(goto_model, options, get_message_handler());
```
Where has this code come from?

smowton committed:
Java frontend: note availability of synthetic methods

This informs symex-driven lazy loading that certain synthetic methods can be
given bodies on demand by the Java frontend, similar to how normal methods we
have bytecode for are treated.

smowton committed:
Fix broken test descriptions

smowton committed:
Java frontend: rename externally driven loading

This name is more accurate, as the frontend is no longer in charge of choosing what gets
loaded and why -- it just obeys convert_lazy_method requests given by the driver program.

smowton committed:
Add callback after symex to bmct

This permits frontend programs to intervene after symex executes, either in addition to
or instead of running bmc. Its current use case is to implement show-goto-functions and
similar debug output options when symex-driven lazy loading means that function bodies are
not populated until symex has run.

When --paths is in use, the callback will be made every time symex finishes a path.

smowton committed:
JBMC: use symex-driven lazy loading

This adds the symex-driven-lazy-loading option, which causes functions to be loaded only when
symex first requests their body. This means that all frontend effort can be saved relating to
functions that symex reveals to be inaccessible, such as virtual function calls whose full array
of possible targets are known to be unreachable as a result of value-set analysis on the type of
the callee instance.

smowton commented:
@tautschnig @peterschrammel @thk123 @mgudemann this is now rebased and ready for review.

tautschnig commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -814,6 +814,9 @@ void java_bytecode_languaget::methods_provided(id_sett &methods) const
   // Add all concrete methods to map
   for(const auto &kv : method_bytecode)
     methods.insert(kv.first);
+  // Add all synthetic methods to map
```
What is a "synthetic method"? Does this in any way relate to stubbing? If so, could we please use a consistent terminology across the code base?

thk123 replied:
I think synthetic methods are things like `clinit_wrapper` that we introduce to handle some JVM mechanisms, I could be wrong though? Perhaps could be defined in, e.g., `synthetic_methods_map.h`?

smowton replied:
Sort of. The synthetic methods are either clinit methods, generated for stub types, or static initializer method wrappers, generated for all static initializers to ensure they only run once, generated whether a stub type or not. In future the methods defined for array types (e.g. `.clone()`) will probably go in this map too. Stub bodies are currently provided with a different mechanism but could use this too.

In general a synthetic method is one generated internally by the language frontend without corresponding user bytecode. Stubs are thus synthetic, but not all synthetics are stubs.

thk123 replied:
I suggest this definition goes somewhere find able in the codebase.

smowton replied:
This isn't really related to this PR, so I've made https://github.com/diffblue/cbmc/pull/1953

tautschnig commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -58,8 +58,8 @@ void java_bytecode_languaget::get_language_options(const cmdlinet &cmd)
     max_user_array_length=std::stoi(cmd.get_value("java-max-vla-length"));
-  if(cmd.isset("lazy-methods-context-sensitive"))
-    lazy_methods_mode=LAZY_METHODS_MODE_CONTEXT_SENSITIVE;
+  if(cmd.isset("symex-driven-lazy-loading"))
```
Somewhere there must be `--help` output describing this, which also needs to be fixed. And so does the parse-options header file.

smowton replied:
Actually no, the option was added ages ago in anticipation of a driver program using it. That is now happening, but the anticipated name isn't the best.

tautschnig replied:
So you're removing an option? Different kind of update, but still one expected?!

smowton replied:
Nobody (except this file) ever accepted it until now. JBMC now accepts "symex-driven-lazy-loading".

tautschnig replied:
Ok.

tautschnig commented on src/cbmc/bmc.cpp:
```diff
@@ -377,6 +377,12 @@ safety_checkert::resultt bmct::execute(
+    // This provides the frontend the opportunity to do things like a
+    // symbol-table or goto-functions dump instead of actually running the
+    // checker, like show-vcc except frontend specific.
+    if(frontend_callback_after_symex && frontend_callback_after_symex())
```
The examples seem strange: why should dumping a symbol table or goto functions require running symex? show-vcc of course does.

thk123 replied:
Because we use symex to load the functions, running show-goto-functions before symex would result in just `__CPROVER_Start`. The comment should probably make it explicit that this is used when looking at the "symbol-table... _in conjunction with symex driven loading_"

smowton replied:
Because with symex-driven lazy loading in play, symex is the only way functions get loaded. Thus show-goto-functions has to come after symex in that case.

smowton replied:
@thk123 added comment.

tautschnig replied:
Sorry, but having to do symbolic execution in order to obtain syntactic information makes no sense. It's ok to forbid lazy loading together with show-goto-functions, but please don't produce a tool that requires a semantic analysis to produce syntactic information.

smowton replied:
This is a reasonable thing to want for debug purposes: if I want to know what got loaded by symex-driven loading, I would pass `--symex-driven-lazy-loading --list-goto-functions` for example. If I want the whole class file I just don't use `symex-driven-lazy-loading`, which processes show/list-goto-functions the old fashioned way.

tautschnig replied:
This implies that the behaviour of show-goto-functions is context-dependent. My take on this is: if you want a way to show the subset of functions that have been loaded during symbolic execution, add a new command-line option. Don't overload an existing one.

smowton replied:
We already have `--lazy-methods`, under which `--show-goto-functions` will again show whatever was loaded, not all the functions in a particular class file.

tautschnig replied:
It didn't claim that was good either ;-)

What I would suggest is additional text to the help output that explains the interaction of the command-line options.

smowton replied:
https://github.com/diffblue/cbmc/pull/1957

tautschnig commented on regression/jbmc-strings/StringIndexOf/test.desc:
```diff
@@ -1,7 +1,14 @@
+see any reason why this is the case except Minisat happening to get down a rabbit hole when
+there are other choices that lead to a quick solution. Disabling the test under symex-driven
+loading for now; most likely it will start working again in the near future for equally
+spurious reasons.
```
Doesn't @romainbrenguier have some fix for indexOf? Is this maybe no longer relevant?

smowton replied:
Re-tested and this indeed appears to have cleared in the meantime. Removed the tag.

thk123 commented on src/cbmc/bmc.h:
```diff
@@ -147,7 +148,8 @@ class bmct:public safety_checkert
     message_handlert &_message_handler,
```
Add doc for the `callback_after_symex` specifically what the return value of the callback should indicate

smowton commented:
TG bump: https://github.com/diffblue/test-gen/pull/1609

thk123 commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -378,6 +377,23 @@ void jbmc_parse_optionst::get_command_line_options(optionst &options)
+  if(cmdline.isset("symex-driven-lazy-loading"))
+  {
+    options.set_option("symex-driven-lazy-loading", true);
+    for(const char *opt :
```
`for(const std::string &opt :`)?

NathanJPhillips replied:
They look like `const char *`'s to me ;-)
If you want to do the construction of a `std::string` for some reason and you want to push it into the for loop then at least use `const std::string` instead of relying on lifetime-extension using `const std::string &` as this will indicate that this is newly created in this scope, not a reference to something created elsewhere.

thk123 commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -508,15 +507,72 @@ int jbmc_parse_optionst::doit()
+        lazy_goto_model,
+        ui_message_handler.get_ui(),
+        *this,
+        configure_bmc,
```
Is this always null here? No - when `java-unwind-enum-static` is set? Perhaps a comment (though I'm wary it risks going out of date).

smowton replied:
Comment added

thk123 commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -865,6 +1003,8 @@ void jbmc_parse_optionst::help()
     // hence its presence here:
     " --java-unwind-enum-static    try to unwind loops in static initialization of enums\n" // NOLINT(*)
+    // Currently only supported in the JBMC frontend:
+    " --symex-driven-lazy-loading  only load functions when first entered by symbolic execution\n" // NOLINT(*)
```
This should be defined in a header file to allow re-use across different driver programs

smowton replied:
Hmmm, it is pretty jbmc-specific at the moment though. I can't see anything except the option name being shareable with test-gen at the moment -- and it surely doesn't belong to bmct or goto_symext, both of which don't care exactly how their get_goto_program function works.

In short I agree some sort of factoring is in order at some point, but I can't see how to do it just now.

tautschnig replied:
But maybe in future CBMC will care?

smowton replied:
This is possible. For now CBMC uses none of the lazy loading infrastructure at all, and I believe @kroening's plan is for it to stay that way for a while. If and when it does, some factoring will be in order.

thk123 commented on regression/cbmc-java/NondetInit/test_lazy.desc:
```diff
@@ -1,4 +1,4 @@
-CORE
+CORE symex-driven-lazy-loading-expected-failure
```
Probably good to add a comment to this `test.desc` explaining why this is an expected failure

thk123 replied:
I see it is because of the `lazy-methods` flag. I still think they'd benefit from a uniform comment, but you may disagree.

allredj replied:
+1 for having a comment on each of the skipped desc to mention the reason they are skipped.

thk123 commented on regression/cbmc-java/lvt-groovy/test.desc:
```diff
@@ -1,7 +1,8 @@
 ^SIGNAL=0$
 --
 --
+This doesn't work under lazy loading because no entry-point function is given.
```
Probably should call it "symex driven lazy loading" since otherwise will get confusing.

thk123 commented on regression/cbmc-java/lvt-unexpected/test.desc:
```diff
@@ -10,3 +10,5 @@ expected by the bytecode to goto conversion algorithms. Namely, the live range
 is perfectly possibly as per the JVM specs but currently additionally required
 in CBMC.
+
+This doesn't work under lazy loading because no entry-point function is given.
```
Probably should call it "symex driven lazy loading" since otherwise will get confusing. et a.

thk123 commented on regression/strings-smoke-tests/CMakeLists.txt:
```diff
@@ -1,3 +1,10 @@
     "$<TARGET_FILE:jbmc>"
 )
+
+add_test_pl_profile(
```
Should these profiles be somehow added to Makefiles as well? Is that even feasible?

smowton replied:
Added

thk123 reviewed (commented):
Nothing blocking - some places where I think a few things could be clearer.

smowton commented:
@tautschnig comments addressed

peterschrammel reviewed (commented):
Re --show-properties: there is a --drop-unused-functions option that also works in combination with  --show-properties. Maybe you can achieve a similar behaviour for --show-properties when used together with your option.

smowton commented:
@thk123 @tautschnig all comments addressed, please re-review

thk123 reviewed (approved):
LGTM (TG bump passes as well).

FWIW my two cents re: `--show-goto-functions` is that the interaction is expected and useful when combined with `--symex-driven-lazy-loading`, it has certainly never surprised me that that is how it behaves for `--lazy-methods`.

I would be disappointed if we lost the ability to show loaded GOTO functions and it seems unnecessarily confusing to have a `--show-loaded-goto-functions` that is only useful when you've got a lazy loading on, and `--show-goto-functions` that's only useful when you've got it off.

smowton committed:
Fix broken static_init_order test descriptions

These were trying to match for a blank line, which doesn't occur in the output
any more, and was surely not intentionally searched for in the first place. This
only causes a problem under Windows (or more likely, under certain versions of Perl),
probably due to a change in empty line stripping.

smowton committed:
Enable show-properties with symex-driven lazy loading

This can be accommodated using the same post-symex reporting phase as show-goto-functions et al.
The one cbmc-java test that uses --show-properties is consequently re-enabled.

smowton commented:
@peterschrammel after a little more thought I realised show-properties could be accommodated similarly to --show-symbol-table et al. I've added a commit switching it back on, along with its test.

NathanJPhillips commented on src/cbmc/bmc.cpp:
```diff
@@ -377,6 +377,15 @@ safety_checkert::resultt bmct::execute(
     const goto_functionst &goto_functions =
       goto_model.get_goto_functions();
 
+    // This provides the frontend the opportunity to do things like a
```
Is this done by a front-end or a driver? If the former then hyphenate, if the latter then reword here and rename the variable.

smowton replied:
Good point, done

NathanJPhillips reviewed (approved):
Might want to squash a couple of the later test commits.

smowton commented:
@tautschnig are you satisfied, can we go ahead?

chrisr-diffblue reviewed (approved):
Reviewing the parts I'm code owner for looks fine by me.
